OutOfSchool/OutOfSchool.Redis/MultiLayerCache.cs

```diff
@@ -0,0 +1,125 @@
+﻿using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.Options;
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace OutOfSchool.Redis;
+
+public sealed class MultiLayerCache : IMultiLayerCacheService, IDisposable
+{
+    private readonly ReaderWriterLockSlim cacheLock = new ReaderWriterLockSlim();
+    private readonly IMemoryCache _memoryCache;
+    private readonly ICacheService _cacheService;
+    private readonly MemoryCacheConfig memoryCacheConfig;
+
+    public MultiLayerCache(
+        ICacheService externalCache,
+        IMemoryCache inMemoryCache,
+        IOptions<MemoryCacheConfig> memoryCacheConfig
+    )
+    {
+        this._memoryCache = inMemoryCache;
+        this._cacheService = externalCache;
+        this.memoryCacheConfig = memoryCacheConfig.Value;
+    }
+
+    public Task ClearCacheAsync(string key)
+    {
+        return ExecuteMultipleAsync(() =>
+        {
+            cacheLock.EnterWriteLock();
+            try
+            {
+                _memoryCache.Remove(key);
+                _cacheService.ClearCacheAsync(key);
+            }
+            finally
+            {
+                cacheLock.ExitWriteLock();
+            }
+        });
+    }
+
+    public async Task<T> GetOrAddAsync<T>(string key, Func<Task<T>> newValueFactory, TimeSpan? absoluteExpirationRelativeToNowInterval = null, TimeSpan? slidingExpirationInterval = null)
```

Comment:
(Logic) For non-existing key I expect `GetOrAddAsync()` and `SetAsync()` to behave similarly. They don't! 
One is asking `_cacheService` to store the data, the other - does not.

Reply:
I deleted Set method from the interface. It doesn`t need in our project.

Sorry, but I can`t catch your mind.
Could you expand it?

Reply:
I'll try to explain
Imagine that you see someone implemented cache interface having two methods - `GetOrAdd` and `Set`. For case of non-existing in cache key you would expect them do the same things. Taking into account that this is `MultiLayerCache` I'd expect that both cache layers will be covered by these Add/Set - both `_cacheService` and `_memoryCache`. `Set` works with `_memoryCache` only.

Is it a good method naming - I'm not sure. 

As now it is not a public method - it's less important, but still not perfect.

Reply:
Done. Deleted Set method.